File: Dynamic-Programming/115. Distinct Subsequences/distinctSubsequences.js
/**
 * @param {string} s
 * @param {string} t
 * @return {number}
 */
 var numDistinct = function(s, t) {
    let m = t.length;
    let n = s.length;
    
    if (m === n && s == t) return 1;
    if (n < m) return 0;
    
    let dp = new Array(m + 1);
    for (let i = 0; i < m + 1; i++) {
        dp[i] = new Array(n + 1).fill(0);
    }
    
    for (let i = 0; i < n + 1; i++) {
        dp[0][i] = 1;
    }
    
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (s[j] === t[i]) {
                dp[i + 1][j + 1] = dp[i][j] + dp[i + 1][j];
            } else {
                dp[i + 1][j + 1] = dp[i + 1][j];
            }
        }
    }
    
    // dp[col][row]
    return dp[m][n]
};

File: Dynamic-Programming/115. Distinct Subsequences/distinctSubsequences.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        m = len(s)
        n = len(t)
        
        # same length and string      
        if m == n and s == t:
            return 1
        # s as the longer string, 
        # length should not be less than t       
        if m < n:
            return 0;
        
        dp = [[0 for _ in range(m + 1)] for _ in range(n + 1)]
        
        # fill in base case - first row
        for i in range(m + 1):
            dp[0][i] = 1
        
        # now looking at each string
        for i in range(n):
            for j in range(m):
                if t[i] == s[j]:
                    dp[i + 1][j + 1] = dp[i][j] + dp[i + 1][j]
                else:
                    dp[i + 1][j + 1] = dp[i + 1][j]
        
        return dp[-1][-1]